p2p/fuzz/src/lib.rs

```diff
@@ -0,0 +1,626 @@
+                                    .insert(from_idx as u8);
+                            }
+                        }
+                    }
+                }
+
+                NetworkOperation::ReceiveMessages => {
+                    // Attempt to receive one message from each receiver with pending messages
+                    let receivers_with_pending: Vec<u8> =
+                        pending_by_receiver.keys().cloned().collect();
+
+                    for to_idx in receivers_with_pending {
+                        let receiver = &mut peers[to_idx as usize].network.receiver;
+
+                        // Try to receive one message with timeout
+                        commonware_macros::select! {
+                            result = receiver.recv() => {
+                                let Ok((sender_pk, message)) = result else {
+                                    continue; // Receive error
+                                };
+
+                                // Identify the sender by their public key
+                                let Some(&from_idx) = topology.pk_to_id.get(&sender_pk) else {
+                                    panic!("Received message from unknown sender: {}", sender_pk);
+                                };
+
+                                // Find the expected queue for this sender-receiver pair
+                                let expected_msgs_key = (to_idx, from_idx);
+                                let Some(queue) = expected_msgs.get_mut(&expected_msgs_key) else {
+                                    panic!("No expected messages for this sender-receiver pair");
+                                };
+
+                                // Find message in expected queue
+                                if let Some(pos) = queue.iter().position(|m| m == &message) {
+                                    // Remove all messages up to and including this one
+                                    // Messages before it were implicitly dropped, this one is received
+                                    for _ in 0..=pos {
+                                        queue.pop_front();
+                                    }
+
+                                    // Clean up empty queue and update pending tracking
+                                    if queue.is_empty() {
+                                        expected_msgs.remove(&expected_msgs_key);
+                                        if let Some(senders) = pending_by_receiver.get_mut(&to_idx) {
+                                            senders.remove(&from_idx);
+                                            if senders.is_empty() {
+                                                pending_by_receiver.remove(&to_idx);
+                                            }
+                                        }
+                                    }
+                                } else {
+                                    panic!(
```

Comment:
See earlier comment about transmission drops

Reply:
In this file we don't enforce that all messages sent from one peer to another are eventually received. However, we _do_ enforce that the messages we do receive are _in order_. If we get to this `panic` then it means that we read a received a `message` but did not find it in the list of messages expected for this sender/receiver pair.